Fetch and stream 3-phase power

daniel-zullo-frequenz opened the pull request:
The 3-phase active power is temporary needed to calculate the power factor of the microgrid. The power factor is not available through [the microgrid API in the current version used in the SDK](https://github.com/frequenz-floss/frequenz-api-common/blob/v0.4.0/proto/frequenz/api/common/v1/metrics/electrical.proto#L139-L199).

So the 3-phase active power is temporary exposed through the microgrid API until the migration to the latest version is
completed. Then the 3-phase power factor can be fetched and streamed through the microgrid API.

Temporary workaround for #841.

daniel-zullo-frequenz committed:
Fetch AC active power per phase from components

Fetch AC active power from meter, inverter and
EV charger components for phase/line 1,2 and 3 respectively.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Add active power per phase metric IDs

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Add active power metrics to data sourcing

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
> So the 3-phase active power is temporary exposed through the microgrid API until the migration to the latest version is
completed. Then the 3-phase power factor can be fetched and streamed through the microgrid API.

I skipped release-notes for that ^ reason

shsms commented on src/frequenz/sdk/timeseries/_power_streamer.py:
````diff
@@ -0,0 +1,197 @@
+        async for active_power_sample in active_power_recv:
+            print(active_power_sample)
+        ```
+    """
````
this is an ambiguous/confusing interface because the 3phase power values are actually for an arbitrary meter, but active power is not a common quantity like frequency or voltage, and can be different in different places in the microgrid.  So `microgrid._active_power()` without a component id or a location in the graph, tells us nothing.

If you stream the power factor instead using power measurements from the arbitrary meter, then that would be a common value for the whole microgrid.

shsms replied:
It should probably also be called `_power_3_phase` if we are going this route.

daniel-zullo-frequenz replied:
For the current use-case we only need the 3-phase power factor. I thought we can compute the 3-phase Power Factor on the application side given that the 3-phase grid current (formula generator) and 3-phase voltage (resampled voltage) are already available in the SDK. So the idea was to create a temporary method `microgrid._active_power()` ("not exposed to the user") to calculate `PF = Active Power / (Voltage . Current)` in the app and drop it once `Power Factor` is available in the SDK
Anyway I think this was just a bad idea given the implications it might cause.

I can try to (1) calculate the Power Factor in the SDK or (2) create the 3-phase power formula generator. The (1) needs to be dropped once the power factor is available in the SDK and (2) could be helpful for another scenarios

daniel-zullo-frequenz replied:
The (3) is just to wait until power factor is available and just drop this PR

shsms replied:
You can't use the grid current for this calculation, unless you're also using the 3-phase grid power.  The current `ActivePowerStreamer` streams the 3-phase power of the `first_descendant_component`, which would be grid power only if the location has a grid-side main meter, which is not everywhere.

shsms replied:
> (1) calculate the Power Factor in the SDK 

for this you could just find the `first_descendant_component` and use raw data from it to do the calculations

> or (2) create the 3-phase power formula generator

I guess with this, you'd make a 3-phase power formula for grid power?

I think both options are fine as a temporary solution, which ever is easier for you.

daniel-zullo-frequenz replied:
The grid.current is 3-phase current, see https://github.com/frequenz-floss/frequenz-sdk-python/blob/v1.x.x/src/frequenz/sdk/timeseries/grid.py#L99-L118

> I guess with this, you'd make a 3-phase power formula for grid power?

Yes, that's what I have in mind.

I'll give a shot to option (2), that should be the easier one :crossed_fingers:

daniel-zullo-frequenz replied:
done, implemented option (2)

shsms reviewed (commented):
Just have concerns about the idea of streaming a 3 phase power for an arbitrary meter, because such a value can't be used in a meaningful way, as active power.  It can only be used for calculating the power factor, in which case we might as well calculate it in the SDK and stream just that.

If we wanted to stream 3 phase power for its own sake, we'd want to create a formula generator for it instead, and create formulas for `grid.power_3_phase`, etc.

I'm also unsure of the idea of streaming resampled power values and using it with unresampled voltage values to calculate power factor.  It would be much better to do it with values coming from a `microgrid.client.meter_data` stream rather than going through the resampler/data sourcing actor.

Also, how does this help calculate power factor if we also don't stream the current values for the arbitrary meter?

daniel-zullo-frequenz committed:
Add tests for 3-phase power metrics

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Add formula engine that streams 3-phase power

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Add formula generator for the grid 3-phase power

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Expose 3-phase power through the grid

The 3-phase power is needed to calculate the power factor
of the microgrid. The power factor is not available through
the microgrid API in the current version used in the SDK.

So the 3-phase power is temporary exposed through the grid
until the migration to the latest version is completed.
Then the 3-phase power factor can be fetched and streamed
through the microgrid.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_grid_power_3_phase_formula.py:
```diff
@@ -0,0 +1,86 @@
+        grid_successors = self._get_grid_component_successors()
+
+        return FormulaEngine3Phase(
+            "grid-3-phase-power",
```
```suggestion
            "grid-power-3-phase",
```

daniel-zullo-frequenz commented:
Renamed `*_3_phase_power` to `*_power_3_phase`

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_grid_power_3_phase_formula.py:
```diff
@@ -0,0 +1,86 @@
+        )
+
+        for idx, comp in enumerate(grid_successors):
+            if idx > 0:
+                formula_builder.push_oper("+")
+
+            # When EV chargers produce `None` samples, they are excluded from
+            # the calculation by treating their `None` values as `0`s.
+            #
+            # This is not possible for Meters, so when they produce `None`
+            # values, those values get propagated as the output.
+            if comp.category in (ComponentCategory.EV_CHARGER,):
+                nones_are_zeros = True
+            elif comp.category == ComponentCategory.METER:
+                nones_are_zeros = False
+            else:
+                continue
```
Don't you want to group inverters with ev chargers?

Also I realized there is a bug in this approach:  If we encounter an unhandled category like inverter, we will `continue`.  But we've already added a `+` step, so we're continuing without adding a component after the `+`.

I think we have to move the `if idx > 0: ...` stuff after the category check,  so that we add a `+` only if we're going to insert a component.

Might affect other formulas as well.

daniel-zullo-frequenz replied:
Makes sense. Should I apply the same patch for [grid current](https://github.com/frequenz-floss/frequenz-sdk-python/blob/v1.x.x/src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_grid_current_formula.py#L55-L72) ?

shsms replied:
yup, I think so.

daniel-zullo-frequenz replied:
> Don't you want to group inverters with ev chargers?

Good catch! Yes, I do want to group them

daniel-zullo-frequenz replied:
Updated as suggested

shsms commented on tests/microgrid/test_component_data.py:
```diff
@@ -53,14 +53,17 @@ def test_inverter_data() -> None:
                     phase_3=electrical_pb2.AC.ACPhase(
                         current=metrics_pb2.Metric(value=34.5),
                         voltage=metrics_pb2.Metric(value=230.2),
+                        power_active=metrics_pb2.Metric(value=3680.1),
```
maybe you want to add an assert for these values as well?

daniel-zullo-frequenz replied:
yes, I do want that. Thanks! Fixed

daniel-zullo-frequenz committed:
Fix grid current formula generation

Add the formula operator if and only if there is a component to add.
Otherwise, this was adding a `+` operator, but not adding a component
after the `+`.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Update bug-fixes in release-notes

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>
